ktx-library-codelab/step-06/app/src/main/java/com/example/android/kotlinktxworkshop/MainActivity.kt

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.android.kotlinktxworkshop
+
+import android.Manifest.permission.ACCESS_FINE_LOCATION
+import android.content.pm.PackageManager
+import android.os.Bundle
+import android.util.Log
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.lifecycleScope
+import com.example.android.myktxlibrary.awaitLastLocation
+import com.example.android.myktxlibrary.findAndSetText
+import com.example.android.myktxlibrary.hasPermission
+import com.example.android.myktxlibrary.locationFlow
+import com.example.android.myktxlibrary.showLocation
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationServices
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.conflate
+
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var fusedLocationClient: FusedLocationProviderClient
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
+
+        lifecycleScope.launchWhenStarted { // only get results when the Activity is in the started state
+            try {
+                getLastKnownLocation()
+                startUpdatingLocation()
+            } catch (e: Exception) {
+                findAndSetText(R.id.textView, "Unable to get location.")
+                Log.d(TAG, "Unable to get location", e)
+            }
+        }
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        if (!hasPermission(ACCESS_FINE_LOCATION)) {
+            requestPermissions(arrayOf(ACCESS_FINE_LOCATION), 0)
+        }
+    }
+
+    private suspend fun getLastKnownLocation() {
+        val lastLocation = fusedLocationClient.awaitLastLocation()
+        showLocation(R.id.textView, lastLocation)
+    }
+
+    private suspend fun startUpdatingLocation() {
+        fusedLocationClient.locationFlow()
+            .conflate()
```

Comment:
This `conflate` operator will still apply backpressure when this Activity is not started is it effectively launches a coroutine in the current flow scope to read & conflate that will be "paused" by the pausing dispatcher.

In this case that's fine due to the implementation of `locationFlow`, however in general this pattern is susceptible to introducing deadlocks to the data layer while in the background as the backpressure effectively creates a circular call graph (the implicit `resume`) and a possible mutex|semaphore.

**Option 1:** Move the conflation coroutine off of the `launchWhenStarted` dispatcher using:

```
locationFlow()
    .flowOn(Dispatchers.Main)
    .conflate()
```

and add a comment about the deadlock concerns.

**Option 2:**  It might be simpler to convert the Flow to a `liveData` before consuming it by the UI, which will do what this code is trying to express in a safer way – and skips over discussions about complex concurrency topics :).

`locationFlow().asLiveData()`

Reply:
Went with option 1 for now, option 2 to come after tomorrow